### PR TITLE
[doc] Add shape for object duplication from Explorer

### LINKED ISSUE #3740: Add support for the duplication of semantic elements in the explorer
An action to duplicate a semantic element should be available in the explorer. Thanks to this action, a dialog should be opened to let the end user select where the object will be duplicated. The target should be valid for such duplication, think for exemple about a mono-valued containment reference which may not be a valid target if it already has a child. The containment feature in the target object should be selectable too. Think for example about an object which has multiple containment references of the same type.

The duplication should be performed in a single action on the backend, nothing should be modified in the clipboard of the end user.

This dialog will should probably let the end user answer some open questions:

- Do we want to duplicate the child elements too?
- Do we want to duplicate outgoing references too ?
- Do we want to copy incoming references too (if they are multi-valued)

### PR DESCRIPTION
**Pull request by lfasani**
# Pull request template

## General purpose
What is the main goal of this pull request?
- [ ] Bug fixes
- [ ] New features
- [x] Documentation
- [ ] Cleanup
- [ ] Tests
- [ ] Build / releng

## Project management
- [x] Has the pull request been added to the relevant project and milestone? (Only if you know that your work is part of a specific iteration such as the current one)
- [x] Have the `priority:` and `pr:` labels been added to the pull request? (In case of doubt, start with the labels `priority: low` and `pr: to review later`)
- [ ] Have the relevant issues been added to the pull request?
- [ ] Have the relevant labels been added to the issues? (`area:`, `difficulty:`, `type:`)
- [ ] Have the relevant issues been added to the same project and milestone as the pull request?
- [ ] Has the `CHANGELOG.adoc` been updated to reference the relevant issues?
- [ ] Have the relevant API breaks been described in the `CHANGELOG.adoc`? (Including changes in the GraphQL API)
- [ ] In case of a change with a visual impact, are there any screenshots in the `CHANGELOG.adoc`? For example in `doc/screenshots/2022.5.0-my-new-feature.png`

## Architectural decision records (ADR)
- [ ] Does the title of the commit contributing the ADR start with `[doc]`?
- [ ] Are the ADRs mentioned in the relevant section of the  `CHANGELOG.adoc`?

## Dependencies
- [ ] Are the new / upgraded dependencies mentioned in the relevant section of the `CHANGELOG.adoc`?
- [ ] Are the new dependencies justified in the `CHANGELOG.adoc`?

## Frontend

This section is not relevant if your contribution does not come with changes to the frontend.

### General purpose
- [ ] Is the code properly tested? (Plain old JavaScript tests for business code and tests based on React Testing Library for the components)

### Typing
We need to improve the typing of our code, as such, we require every contribution to come with proper TypeScript typing for both changes contributing new files and those modifying existing files.
Please ensure that the following statements are true for each file created or modified (this may require you to improve code outside of your contribution).

- [ ] Variables have a proper type
- [ ] Functions’ arguments have a proper type
- [ ] Functions’ return type are specified
- Hooks are properly typed:
	- [ ] `useMutation<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useQuery<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useSubscription<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useMachine<CONTEXT_TYPE, EVENTS_TYPE>(…)`
	- [ ] `useState<STATE_TYPE>(…)`
- [ ] All components have a proper typing for their props
- [ ] No useless optional chaining with `?.` (if the GraphQL API specifies that a field cannot be `null`, do not treat it has potentially `null` for example)
- [ ] Nullable values have a proper type (for example `let diagram: Diagram | null = null;`)

## Backend

This section is not relevant if your contribution does not come with changes to the backend.

### General purpose
- [ ] Are all the event handlers tested?
- [ ] Are the event processor tested?
- [ ] Is the business code (services) tested?
- [ ] Are diagram layout changes tested?

### Architecture
- [ ] Are data structure classes properly separated from behavioral classes?
- [ ] Are all the relevant fields final?
- [ ] Is any data structure mutable? If so, please write a comment indicating why
- [ ] Are behavioral classes either stateless or side effect free?

### How to test this PR?
_Please describe here the various use cases to test this pull request_

- [ ] Has the Kiwi TCMS test suite been updated with tests for this contribution?

### CODE REVIEW

**Review by sbegaudeau — COMMENTED** [review]
Please only use the labels `pr: xxx` and `priority: xxx` for pull requests

File: doc/iterations/2024.9/add_support_for_copy_paste.adoc [hunk 1]
```diff
@@ -0,0 +1,93 @@
+= Add support for copy paste
```
Comment (sbegaudeau) [inline doc/iterations/2024.9/add_support_for_copy_paste.adoc]: None of this work is related to an action to copy nor the action of pasting at all. It is explicitly the exclusion of both of these use cases that has motivated the inclusion of this task in this iteration. The only technical perimeter that we agreed on considering for this iteration is an action to duplicate semantic elements.

The major difference being that the duplication is instantaneous. In the perimeter of this work, one cannot copy a semantic element from Sirius Web and paste it elsewhere (Microsoft Word, an email, another tab of Sirius Web, etc) nor copying something from outside (Microsoft Excel for example) and paste it in Sirius Web.

Copying and pasting bring significant technical challenges that we decided not to tackle on purpose. Duplicating is hard enough to perform. Here, there will be one action (duplicating) which will copy and paste at once.

Nothing here, should mention copying or pasting, this is not the function nor technical perimeter considered for this iteration.

File: doc/iterations/2024.9/add_support_for_copy_paste.adoc [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+== Problem
+
+Currently, it is not possible to copy and paste something in Sirius Web.
```
Comment (sbegaudeau) [inline doc/iterations/2024.9/add_support_for_copy_paste.adoc]: duplicate

File: doc/iterations/2024.9/add_support_for_copy_paste.adoc [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+== Key Result
+
+The user can copy a semantic object with a shortcut or with contextual menu in the explorer.
+The user can paste on an other semantic object.
+The effect is that the copied object is duplicated into the object where it is pasted.
```
Comment (sbegaudeau) [inline doc/iterations/2024.9/add_support_for_copy_paste.adoc]: Those are some of the various use cases that have been explicitly excluded from this work.

File: doc/iterations/2024.9/add_support_for_copy_paste.adoc [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+When executing `Copy` the copied element is the semantic object related to the selected object.
+For example, copying
+* an item in the Explorer Tree will copy the associated semantic element
+* a node in a diagram will copy the semantic element associated to the node
+
+To know the selected semantic element, see what is displayed in the `Details` view.
+
+This is the same logic for the `Paste`. The copied object is pasted into the selected element.
+For example:
+* a document in the Explorer Tree will paste into the document
```
Comment (sbegaudeau) [inline doc/iterations/2024.9/add_support_for_copy_paste.adoc]: In a first version, given the appetite we have for this feature, only semantic elements based on EMF from the explorer can be duplicated, using a dedicated dialog showing only the valid locations where the object can be pasted.

File: doc/iterations/2024.9/add_support_for_copy_paste.adoc [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+
+This behavior described below is the default implementation.
+The backend implementation will rely on spring bean contribution that will allow to provide custom implementations.
+
+*How to copy and paste?*
+
+The shortcut `Ctrl + C` for copy and `Ctrl + V` for paste are available anywhere.
+
+`Copy` action is added in contextual menu on tree items in `Explorer` except for representations and models.
+`Paste` action is added in contextual menu on tree items in `Explorer` except for representations.
```
Comment (sbegaudeau) [inline doc/iterations/2024.9/add_support_for_copy_paste.adoc]: No shortcut will be available for that, this is not about copying and pasting

**Review by sbegaudeau — CHANGES_REQUESTED** [review]
I've just reviewed the first part of the document but it does not match what was discussed.

File: doc/iterations/2024.9/add_support_for_copy_paste.adoc (hunk 1 above)
Comment (lfasani) [reply doc/iterations/2024.9/add_support_for_copy_paste.adoc]: The shape has been completely updated for fulfilling the duplicate specification

File: doc/iterations/2024.9/add_support_for_object_duplication.adoc [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+
+The aim of `Duplicate` is to duplicate the selected semantic object into a selected container given some options.
+
+When clicking on `Duplicate`, it opens a dialog which allows the user to choose how the duplication will be done.
+
+image:./images/duplicate_object_dialog.png[duplicate object dialog]
+
+* `Container` part allow to define the container of the duplicated object.
+It presents a tree that is filtering so that only valid target container are displayed.
+A valid target container is an object that owns at least one multi-valued containment feature that can contain the object about to be duplicated.
```
Comment (frouene) [inline doc/iterations/2024.9/add_support_for_object_duplication.adoc]: Only works for multi-valued feature ?
It  won't be possible to duplicate an object to another mono-valued feature ?

Comment (lfasani) [reply doc/iterations/2024.9/add_support_for_object_duplication.adoc]: yes, we could accept containment mono-valued feature if not set with a non-null value.

**Review by sbegaudeau — APPROVED** [review]
👍

**Pull request merged**